### ISSUE TITLE
Implement logging + talk to team 2 about it

### ISSUE THREAD

**Issue by Erdnaf** [privML/privacy-evaluator#153]
For better user experience, while waiting for attacks to finish, provide some basic per step logging.

**Comment by blauertee** [comment]
Hey team2 as I see you added some verbose stuff in #148. We wanted to do that too in this sprint and we think it would be best practice to use the `logging` module instead of `print` statements. Would you be willing to change that? 
For you that would just mean to replace `print(x)` by `logging.debug(x)` or `logging.warn(x)`.

If you wanna discuss this in detail we can meet on Tuesday again :)

**Comment by yzchyx** [comment]
Hey, @blauertee . We also thought about the `logging` module since it is already built-in and provides 5-level-logger natively. Besides we can set logLevel everywhere to make it look like global variable and change that by parameter of a function if necessary, so in general it is quite flexible.
However the thing is, we can not find a clean and simple way to make `logging` and `tqdm` compatible with each other.
Another issue is, the naming of the levels provided by `logging` is kinda confusing in this case because wanting more detailed information from the program has nothing to do with `error` or something like that.
In summary, we wanted to implement all the logger with `logging` but had to give up mainly because of the two issues which can not be elegantly solved based on our knowledge.

**Comment by blauertee** [comment]
What I got from looking at tqdm for a few minutes is that should always use `tdqm.write` instead of `print` so the same error should arise with the `print` statements. Can you quickly explain where the difference to logging is? 
Ans also it seems pretty simple to redirect logging to `tqdm.write` as explained [here](https://github.com/tqdm/tqdm#redirecting-logging).

**Comment by yzchyx** [comment]
Thanks for your quick reply!

As to the second thing, since Franziska told us to do 3-level verbose mode: 0: no info at all, 1: important info, 2: most detailed info. And we look at the logging levels, and think level 2 has nothing to do with either "warning", "error" or "critical". We know it's just a name, but since it is a notebook provided to non-experts, they might feel confused that they "make something wrong" after just going level 2 for "detailed info" as we claimed to provide, and then seeing such outputs. 
The first thing is the mechanism of `tqdm` because it occupies the `stderr`-channel by default, which enjoys higher priority than `stdout` (e.g. print). this mechanism causes problem that the order of display is not consistent with the order we put them in the program. Of course we don't want to flush every time before calling tqdm, so we need to redirect by either the way you gave a link to me or the way we did in our branch (manually specify the direction). But personally I prefer `logging`'s fashion and don't think it is quite that ecky.

**Comment by blauertee** [comment]
To the First thing I'd say you can use debug and Info, because you dont need to print anything for "No Info at all" right? And critical info should even be displayed when the user suppresses output I'd say, if users really wanna suppress any output they can still pipe it to `/dev/null`.
It sounds like you otherwise don't have any strong objections to use logging, so maybe just communicate back to your team that we'd be in favor of using logging and if that's fine with everyone and does not cause any bigger issues, let's do it :3

**Comment by yzchyx** [comment]
Thanks for your reply!
Actually I myself tended to use the `logging` but we had to fulfill the sprint backlog first.

I think it is just different topics between our two groups. When we had our sprint meeting Franziska wanted us to add the "verbose" argument just like in `sklearn` which is involved in all attack functions and is part of the `*args` or `**kwargs` optional parameters. Since such requirements #137 and #136 are defined in sprint backlog this way, we studied the source code and implemented using `verbose` variable to control printing as she wanted. If we chose to do it in `logging` we would fail #136 too because she wanted a verbose variable to be "hidden" in asterisk most of the time.

I will ask Franziska next meeting if we could really do it in `logging` and if it is okay we don't apply  verbose fashion, hope it works.

**Comment by blauertee** [comment]
- [ ] add to readme
- [ ] add to notebooks
